File: HW5/tetromino.cpp
#include "tetromino.hpp"
#include "vector.hpp"

/* I created this function to reach number_of_row.. */
int Tetromino :: get_row() const
{
    return (number_of_row);
}

/* I created this function to reach number of column.. */
int Tetromino :: get_column() const
{
    return (number_of_column);
}


/* I created this function to reach indexes I want of shape array... */
char Tetromino :: get_value_of_index(int i, int j) const
{
    return (shape[i][j]);
}


char Tetromino :: get_hold_enum() const
{
    return (hold_enum);
}


Tetromino :: Tetromino(NumberOfShape choice)
{
    VariableOfShape = (int) choice; // I assigned my enum variable to an integer variable. My purpose is that can use it in next lines...
    if (choice == NumberOfShape::I)  // I filled my arrays.
    {
        shape.push_back ({'I', 'I', 'I', 'I'});
        hold_enum = 'I';
        number_of_row = 1;
        number_of_column = 4;
    }
    else if (choice == NumberOfShape::O)
    {
        shape.push_back ({'O', 'O'});
        shape.push_back ({'O', 'O'});
        hold_enum = 'O';
        number_of_row = 2;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::T)
    {
        shape.push_back ({'T', 'T', 'T'});
        shape.push_back ({' ', 'T', ' '});
        hold_enum = 'T';
        number_of_row = 2;
        number_of_column = 3;
    }
    else if (choice == NumberOfShape::J)
    {
        shape.push_back ({' ', 'J'});
        shape.push_back ({' ', 'J'});
        shape.push_back ({'J', 'J'});
        hold_enum = 'J';
        number_of_row = 3;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::L)
    {
        shape.push_back ({'L', ' '});
        shape.push_back ({'L', ' '});
        shape.push_back ({'L', 'L'});
        hold_enum = 'L';
        number_of_row = 3;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::S)
    {
        shape.push_back ({' ', 'S', 'S'});
        shape.push_back ({'S', 'S', ' '});
        hold_enum = 'S';
        number_of_row = 2;
        number_of_column = 3;
    }
    else if (choice == NumberOfShape::Z)
    {
        shape.push_back ({'Z', 'Z', ' '});
        shape.push_back ({' ', 'Z', 'Z'});
        hold_enum = 'Z';
        number_of_row = 2;
        number_of_column = 3;
    }
}


Tetromino :: Tetromino()  // I created this constructor in my main file first to be able to create the objects of this class without using the constructor parameters.
{}


void Tetromino :: print() const  // I created this function to print my array..
{
    for (int i = 0; i < number_of_row; i++)  // number of rows
    {
        for (int j = 0; j < number_of_column; j++)  // number of colummns
        {
            cout << shape[i][j];  //  I printed in here...
        }
        cout << endl;
    }
    cout << endl;
}



void Tetromino :: rotate(char direction_of_rotation)
{
    vector < vector < char > > temp(number_of_column);  // I created temporary array in here..
    int temp_int;
    
    for (int i = 0; i < number_of_column; i++)   // By relocating the dimensions of my original vector, I defined the temporal vector as large enough to hold the return values. Shape's row -> temp's column, Shape's column -> temp's row
    {
        for(int j = 0; j < number_of_row; j++)
        {
            temp[i].push_back(' ');
        }
    }

    if (VariableOfShape != 1) // I never rotated 'O' because it will take the same shape when it is rotated.
    {
        if (direction_of_rotation == 'r')  // 'r' means turn to the right side..
        {
            if (VariableOfShape == 0)  // I sent the values ​​in the shape vector to the temp vector according to the coordinates that will occur when it returns...
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'I')
                            temp[j][number_of_row - 1 - i] = 'I';
            }
            else if (VariableOfShape == 2)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'T')
                            temp[j][number_of_row - 1 - i] = 'T';
            }
            else if (VariableOfShape == 3)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'J')
                            temp[j][number_of_row - 1 - i] = 'J';
            }
            else if (VariableOfShape == 4)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'L')
                            temp[j][number_of_row - 1 - i] = 'L';
            }
            else if (VariableOfShape == 5)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'S')
                            temp[j][number_of_row - 1 - i] = 'S';
            }
            else if (VariableOfShape == 6)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'Z')
                            temp[j][number_of_row - 1 - i] = 'Z';
            }
            shape = temp;    // I equated the vector named temp that I created to the vector named shape..
        }
        else      //  means turn to the left side..
        {
            if (VariableOfShape == 0)  // I sent the values ​​in the shape vector to the temp vector according to the coordinates that will occur when it returns...
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'I')
                            temp[number_of_column - 1 - j][i] = 'I';
            }
            else if (VariableOfShape == 2)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'T')
                            temp[number_of_column - 1 - j][i] = 'T';
            }
            else if (VariableOfShape == 3)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'J')
                            temp[number_of_column - 1 - j][i] = 'J';
            }
            else if (VariableOfShape == 4)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'L')
                            temp[number_of_column - 1 - j][i] = 'L';
            }
            else if (VariableOfShape == 5)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'S')
                            temp[number_of_column - 1 - j][i] = 'S';
            }
            else if (VariableOfShape == 6)
            {
                for (int i = 0; i < number_of_row; i++)
                    for (int j = 0; j < number_of_column; j++)
                        if (shape[i][j] == 'Z')
                            temp[number_of_column - 1 - j][i] = 'Z';
            }
            shape = temp;     // I equated the vector named temp that I created to the vector named shape..
        }
    }

    temp_int = number_of_row;
    number_of_row = number_of_column;         // Since the column and row numbers will change, I saved it again after returning it.
    number_of_column = temp_int;       // Since the column and row numbers will change, I saved it again after returning it.
}



vector < vector < char > > Tetromino :: return_shape()
{
    return (shape);
}


/* I created this operator overloading to create new tetris which has new dimensions...*/
/* This function is same with my constructor...*/
void Tetromino :: operator()(NumberOfShape choice)
{
    VariableOfShape = (int) choice; // I assigned my enum variable to an integer variable. My purpose is that can use it in next lines...
    if (choice == NumberOfShape::I)  // I filled my arrays.
    {
        shape.push_back ({'I', 'I', 'I', 'I'});
        hold_enum = 'I';
        number_of_row = 1;
        number_of_column = 4;
    }
    else if (choice == NumberOfShape::O)
    {
        shape.push_back ({'O', 'O'});
        shape.push_back ({'O', 'O'});
        hold_enum = 'O';
        number_of_row = 2;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::T)
    {
        shape.push_back ({'T', 'T', 'T'});
        shape.push_back ({' ', 'T', ' '});
        hold_enum = 'T';
        number_of_row = 2;
        number_of_column = 3;
    }
    else if (choice == NumberOfShape::J)
    {
        shape.push_back ({' ', 'J'});
        shape.push_back ({' ', 'J'});
        shape.push_back ({'J', 'J'});
        hold_enum = 'J';
        number_of_row = 3;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::L)
    {
        shape.push_back ({'L', ' '});
        shape.push_back ({'L', ' '});
        shape.push_back ({'L', 'L'});
        hold_enum = 'L';
        number_of_row = 3;
        number_of_column = 2;
    }
    else if (choice == NumberOfShape::S)
    {
        shape.push_back ({' ', 'S', 'S'});
        shape.push_back ({'S', 'S', ' '});
        hold_enum = 'S';
        number_of_row = 2;
        number_of_column = 3;
    }
    else if (choice == NumberOfShape::Z)
    {  
        shape.push_back ({'Z', 'Z', ' '});
        shape.push_back ({' ', 'Z', 'Z'});
        hold_enum = 'Z';
        number_of_row = 2;
        number_of_column = 3;
    }
}


/* This is my destructor...*/
Tetromino :: ~Tetromino() {}

File: HW7/Tetromino.java
/**
 *
 *@author Ahmet Furkan Ekinci - 200104004063
 *
 */

import java.util.Scanner;

    /**
     *
     * This class is Tetris class.
     *
     */
    public class Tetromino
    {
        private char[][] shape;
        private NumberOfShape variableOfShape;
        private int numberOfColumn;
        private int numberOfRow;
        /**
         * Enumeration class representing the different shapes of Tetromino
         * @author Ahmet Furkan Ekinci
         */
        public enum NumberOfShape
        {
            /**
             * The I shape, represented by a vertical line of 4 blocks
             */
            I,
            /**
             * The O shape, represented by a 2x2 square of blocks
             */
            O,
            /**
             * The T shape, represented by a T-like shape of 3 blocks
             */
            T,
            /**
             * The J shape, represented by a L-like shape of 3 blocks
             */
            J,
            /**
             * The L shape, represented by a reversed L-like shape of 3 blocks
             */
            L,
            /**
             * The S shape, represented by a Z-like shape of 3 blocks
             */
            S,
            /**
             * The Z shape, represented by a reversed Z-like shape of 3 blocks
             */
            Z
        }


        
        /**
         * Default constructor for creating Tetromino objects without parameters.
         */
        public Tetromino(){/* I created this constructor in my main file first to be able to create the objects of this class without using the constructor parameters.*/}
        /**
         * Constructor for creating Tetromino objects with a specific shape.
         * @param choice is an NumberOfShape parameter, represents the shape of the Tetromino object.
         */
        public Tetromino(NumberOfShape choice)
        {
            variableOfShape = choice; // I assigned my enum variable to an integer variable. My purpose is that can use it in next lines...
            if (choice == NumberOfShape.I)  // I filled my arrays.
            {
                shape = new char[1][4];
                for (int i = 0; i < 4; i++)
                {
                    shape[0][i] = 'I';
                }
                numberOfRow = 1;
                numberOfColumn = 4;
            }
            else if (choice == NumberOfShape.O)
            {
                shape = new char[2][2];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][0] = 'O';
                    shape[i][1] = 'O';
                }
                numberOfRow = 2;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.T)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'T';
                }
                shape[0][0] = 'T';
                shape[0][2] = 'T';
                shape[1][0] = ' ';
                shape[1][2] = ' ';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
            else if (choice == NumberOfShape.J)
            {
                shape = new char[3][2];
                for (int i = 0; i < 3; i++)
                {
                    shape[i][1] = 'J';
                }
                shape[0][0] = ' ';
                shape[1][0] = ' ';
                shape[2][0] = 'J';
                numberOfRow = 3;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.L)
            {
                shape = new char[3][2];
                for (int i = 0; i < 3; i++)
                {
                    shape[i][0] = 'L';
                }
                shape[0][1] = ' ';
                shape[1][1] = ' ';
                shape[2][1] = 'L';
                numberOfRow = 3;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.S)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'S';
                }
                shape[0][0] = ' ';
                shape[0][2] = 'S';
                shape[1][0] = 'S';
                shape[1][2] = ' ';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
            else if (choice == NumberOfShape.Z)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'Z';
                }
                shape[0][0] = 'Z';
                shape[0][2] = ' ';
                shape[1][0] = ' ';
                shape[1][2] = 'Z';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
        }
        /**
         *This method is getter for row.
         *
         *@return The value length of row for Tetromino object
         *
         */
        public int getRow()
        {
            return (numberOfRow);
        }
        /**
         *This method is getter for row.
         *
         *@return The value length of column for Tetromino object
         *
         */
        public int getColumn()
        {
            return (numberOfColumn);
        }
        /**
         *This is actually like my one parameter constructor because their aim is same but it is not constructor.
         *
         *@param choice NumberOfShape parameter used in the method..
         *
         */
        public void bracket(NumberOfShape choice) // parantez overloadının yerine kullanacağğım
        {
            variableOfShape = choice; // I assigned my enum variable to an integer variable. My purpose is that can use it in next lines...
            if (choice == NumberOfShape.I)  // I filled my arrays.
            {
                shape = new char[1][4];
                for (int i = 0; i < 4; i++)
                {
                    shape[0][i] = 'I';
                }
                numberOfRow = 1;
                numberOfColumn = 4;
            }
            else if (choice == NumberOfShape.O)
            {
                shape = new char[2][2];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][0] = 'O';
                    shape[i][1] = 'O';
                }
                numberOfRow = 2;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.T)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'T';
                }
                shape[0][0] = 'T';
                shape[0][2] = 'T';
                shape[1][0] = ' ';
                shape[1][2] = ' ';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
            else if (choice == NumberOfShape.J)
            {
                shape = new char[3][2];
                for (int i = 0; i < 3; i++)
                {
                    shape[i][1] = 'J';
                }
                shape[0][0] = ' ';
                shape[1][0] = ' ';
                shape[2][0] = 'J';
                numberOfRow = 3;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.L)
            {
                shape = new char[3][2];
                for (int i = 0; i < 3; i++)
                {
                    shape[i][0] = 'L';
                }
                shape[0][1] = ' ';
                shape[1][1] = ' ';
                shape[2][1] = 'L';
                numberOfRow = 3;
                numberOfColumn = 2;
            }
            else if (choice == NumberOfShape.S)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'S';
                }
                shape[0][0] = ' ';
                shape[0][2] = 'S';
                shape[1][0] = 'S';
                shape[1][2] = ' ';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
            else if (choice == NumberOfShape.Z)
            {
                shape = new char[2][3];
                for (int i = 0; i < 2; i++)
                {
                    shape[i][1] = 'Z';
                }
                shape[0][0] = 'Z';
                shape[0][2] = ' ';
                shape[1][0] = ' ';
                shape[1][2] = 'Z';
                numberOfRow = 2;
                numberOfColumn = 3;
            }
        }
        /**
         *This method is getter for an index in shape variable of object.
         *
         *@param i is a integer parameter. It is about index...
         *@param j is a integer parameter. It is about index...
         *
         *@return The result of the task as a char value. It means a value in the address whisch we want.
         *
         */
        public char getValueOfIndex(int i, int j)
        {
            return (shape[i][j]);
        }
        /**
         *
         * This method for printing object of Tetromino.
         *
         */
        public void print()
        {
            for (int i = 0; i < numberOfRow; i++)  // number of rows
            {
                for (int j = 0; j < numberOfColumn; j++)  // number of colummns
                {
                    System.out.print(shape[i][j]);  //  I printed in here...
                }
                System.out.print("\n");
            }
            System.out.print("\n");
        }
        /**
         *This method to rotate shape variable of object.
         *
         *@param directionOfRotation is a char parameter in this method.
         *
         */
        public void rotate(char directionOfRotation)
        {
            char[][] temp;  // I created temporary array in here..
            int tempInt;
            
            temp = new char[numberOfColumn][numberOfRow];
            
            for (int i = 0; i < numberOfColumn; i++)   // By relocating the dimensions of my original vector, I defined the temporal vector as large enough to hold the return values. Shape's row -> temp's column, Shape's column -> temp's row
            {
                for(int j = 0; j < numberOfRow; j++)
                {
                    temp[i][j] = ' ';
                }
            }
            if (variableOfShape != NumberOfShape.O) // I never rotated 'O' because it will take the same shape when it is rotated.
            {
                if (directionOfRotation == 'r')  // 'r' means turn to the right side..
                {
                    if (variableOfShape == NumberOfShape.I)  // I sent the values ​​in the shape vector to the temp vector according to the coordinates that will occur when it returns...
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'I')
                                    temp[j][numberOfRow - 1 - i] = 'I';
                    }
                    else if (variableOfShape == NumberOfShape.T)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'T')
                                    temp[j][numberOfRow - 1 - i] = 'T';
                    }
                    else if (variableOfShape == NumberOfShape.J)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'J')
                                    temp[j][numberOfRow - 1 - i] = 'J';
                    }
                    else if (variableOfShape == NumberOfShape.L)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'L')
                                    temp[j][numberOfRow - 1 - i] = 'L';
                    }
                    else if (variableOfShape == NumberOfShape.S)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'S')
                                    temp[j][numberOfRow - 1 - i] = 'S';
                    }
                    else if (variableOfShape == NumberOfShape.Z)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'Z')
                                    temp[j][numberOfRow - 1 - i] = 'Z';
                    }
                    shape = new char[numberOfColumn][numberOfRow];
                    for (int i = 0; i < numberOfColumn; i++)
                        for (int j = 0; j < numberOfRow; j++)
                            shape[i][j] = temp[i][j];    // I equated the vector named temp that I created to the vector named shape..
                }
                else      //  means turn to the left side..
                {
                    if (variableOfShape == NumberOfShape.I)  // I sent the values ​​in the shape vector to the temp vector according to the coordinates that will occur when it returns...
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'I')
                                    temp[numberOfColumn - 1 - j][i] = 'I';
                    }
                    else if (variableOfShape == NumberOfShape.T)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'T')
                                    temp[numberOfColumn - 1 - j][i] = 'T';
                    }
                    else if (variableOfShape == NumberOfShape.J)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'J')
                                    temp[numberOfColumn - 1 - j][i] = 'J';
                    }
                    else if (variableOfShape == NumberOfShape.L)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'L')
                                    temp[numberOfColumn - 1 - j][i] = 'L';
                    }
                    else if (variableOfShape == NumberOfShape.S)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'S')
                                    temp[numberOfColumn - 1 - j][i] = 'S';
                    }
                    else if (variableOfShape == NumberOfShape.Z)
                    {
                        for (int i = 0; i < numberOfRow; i++)
                            for (int j = 0; j < numberOfColumn; j++)
                                if (shape[i][j] == 'Z')
                                    temp[numberOfColumn - 1 - j][i] = 'Z';
                    }
                    shape = new char[numberOfColumn][numberOfRow];
                    for (int i = 0; i < numberOfColumn; i++)
                        for (int j = 0; j < numberOfRow; j++)
                            shape[i][j] = temp[i][j];    // I equated the vector named temp that I created to the vector named shape..
                }
            }

            tempInt = numberOfRow;
            numberOfRow = numberOfColumn;         // Since the column and row numbers will change, I saved it again after returning it.
            numberOfColumn = tempInt;       // Since the column and row numbers will change, I saved it again after returning it.
        }
        /**
         *This method is getter for shape variable of object.
         *
         *@return This method return char[][]...
         *
         */
        public char[][] returnShape()
        {
            return (shape);
        }
    
    }

